### ISSUE TITLE
target_tensor argument from tf.keras.Model compile() method has dissapeared

### ISSUE THREAD

**Issue by scarlehoff** [tensorflow/tensorflow#41248]
I am not entirely sure whether this is a bug or not.

I have a piece of code which was relying in the `target_tensors` argument of the `compile` method. This code works fine in TF 2.1, however when updating to TF 2.2 it stopped working.
Looking in the documentation I've noticed that indeed, the argument `target_tensors` has dissapeared [link](https://www.tensorflow.org/api_docs/python/tf/keras/Model#compile) however I haven't seen any mentions to this in the changelog.

Is this a bug or is it intentional? The error message just says
`
ValueError: target_tensors argument is not supported when executing eagerly
`

But looking at the source code https://github.com/tensorflow/tensorflow/blob/ae975ee7445b448abb40a8916336dc5646e8b46a/tensorflow/python/keras/engine/training.py#L2496 this argument is clearly not accepted anymore.

If it is intentional, is there a recommended workaround? I can think of several not-very-intrusive ways around it, but all them feel a bit hacky.

**Comment by amahendrakar** [comment]
@scarlehoff,
In order to expedite the trouble-shooting process, could you please provide the complete code to reproduce the issue reported here. Thanks!

**Comment by scarlehoff** [comment]
Hi @amahendrakar sadly I cannot provide the code, but I've written a MWE that works fine in TF 2.1 but not in TF 2.2 or 2.3
```

import tensorflow as tf
import numpy as np

x = np.random.rand(1, 10)
y = np.random.rand(4)

my_model = tf.keras.Sequential(layers=tf.keras.layers.Dense(4, input_shape=(10,) ))
my_model.compile(loss='mse', target_tensors=[y], run_eagerly=False)

my_model.fit(x)
```

In TF 2.1 works while in 2.2 I get the following error:

`ValueError: target_tensors argument is not supported when executing eagerly.`

Note that I am explicitly setting `run_eagerly=False` (even thought I believe that's the default).
I know I can use `disable_eager_execution` to change the behaviour of my whole codebase but that's not what I would like to do.

In short, I would like to know whether this is an intentional change (i.e., `target_tensors` is basically not supported any more, since it is not included in the documentation).

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.2](https://colab.research.google.com/gist/amahendrakar/20fbce37e51bd27e2a57ea0e1a949316/41248.ipynb#scrollTo=wW5LjcAYgtyh), [TF v2.3.0-rc1](https://colab.research.google.com/gist/amahendrakar/a10f3be1d3b184b0359ae7ff01ca0665/41248-2-3.ipynb#scrollTo=vH4tBjkZhAOO) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/74ee054b72deb740aa8cb8755deb4235/41248-tf-nightly.ipynb). Facing an error stating `ValueError: target_tensors argument is not supported when executing eagerly.` on running the code. 

However, not facing any issues while running with [TF v2.1](https://colab.research.google.com/gist/amahendrakar/44cefdb88221ca255e4f02d0006bde7b/41248-2-1.ipynb). Please find the attached gist. Thanks!

**Comment by omalleyt12** [comment]
@scarlehoff Thanks for the issue!

Yes, `target_tensors` is no longer supported, as internally we shifted Keras away from using a global graph and towards using `tf.function`, and so the target-tensors concept was no longer feasible to support. Could you advise on a minimal example of what you'd like to use target tensors for, and I can advise on a workaround?

**Comment by scarlehoff** [comment]
Hi @omalleyt12, one example is the one I wrote above. I've already written a workaround for my particular problem (that is as fast as it was with 2.1, the main reason I was using `target_tensors` is speed). 

It would be really helpful to change the error message when using `target_tensors` as it is a red herring (my first thought was, oh, no, somehow everything is running in eager mode now!)

In any case, even if I've already solved my problem let me include a minimal version of it and its workaround because surely there is a better way. It could also be helpful for some people looking for a solution, even if happens to be suboptimal.

To first approximation would like to be able to do this:
```
import tensorflow as tf
import numpy as np

x = np.random.rand(1, 10)
y = np.random.rand(4)

my_model = tf.keras.Sequential(layers=tf.keras.layers.Dense(4, input_shape=(10,) ))
my_model.compile(loss='mse', target_tensors=[y], run_eagerly=False)

for _ in range(epochs):
    my_model.evaluate(x, y = None)
```

Since passing the y values for every step makes it become really slow.
```
for _ in range(epochs):
    my_model.evaluate(x, y = y)
```

My solution has been to use `make_test_function` to create a custom `evaluate` function which has already the `y` value compiled like

```
    def make_test_function(self):
        @tf.function
        def evaluate():
            ret = calculate_loss_of_model()
            return ret
        return evaluate
```
so I can indeed do `my_model.evaluate(x, y = None)` and recover the same performance I had when using TF 2.1.
Please let me know if you think there's a better way.

**Comment by omalleyt12** [comment]
Looks good to me, the other thing is that usually `model.evaluate` should be called on all the data at once, it's a batch-processing API. Calling it for each batch will be slow bc the `tf.function` might retrace if it can't determine that the input data is of the same type, shape, dtype, etc
